packages/datatrak-web/src/database/entity/getEntityDescendants.ts

```diff
@@ -0,0 +1,226 @@
+import { Project } from '@tupaia/types';
+
+const DEFAULT_FIELDS = ['id', 'parent_name', 'code', 'name', 'type'] as ExtendedEntityFieldName[];
+
+const DEFAULT_PAGE_SIZE = 100;
+
+type SearchResult = {
+  name: string;
+  parent?: {
+    name: string;
+  };
+};
+
+export type GetEntityDescendantsParams = {
+  filter?: Record<string, unknown>;
+  fields?: ExtendedEntityFieldName[];
+  pageSize?: number;
+  searchString?: string;
+};
+
+export function sortSearchResults(searchString: string, results: SearchResult[]) {
+  const lowerSearch = searchString.toLowerCase();
+
+  const primarySearchResults = results.filter(({ name }) =>
+    name.toLowerCase().startsWith(lowerSearch),
+  );
+
+  const secondarySearchResults = results.filter(
+    ({ name, parent }) =>
+      !name.toLowerCase().startsWith(lowerSearch) &&
+      (name.toLowerCase().includes(lowerSearch) ||
+        parent?.name.toLowerCase().startsWith(lowerSearch)),
+  );
+
+  return [...primarySearchResults, ...secondarySearchResults];
+}
+
+const getAllowedCountries = async (
+  models: DatatrakWebModelRegistry,
+  rootEntityId: string,
+  project: ProjectRecord,
+  isPublic: boolean,
+  accessPolicy: AccessPolicy,
+) => {
+  const rootEntity = await models.entity.findById(rootEntityId);
+
+  if (!project.entity_hierarchy_id) {
+    throw new Error('Project entity hierarchy ID is not set');
+  }
+
+  const countryEntities = await rootEntity.getChildrenFromParentChildRelation(
+    project.entity_hierarchy_id,
+  );
+
+  let allowedCountries = countryEntities
+    .map(child => child.country_code)
+    .filter(isNotNullish)
+    .filter((countryCode, index, countryCodes) => countryCodes.indexOf(countryCode) === index); // De-duplicate countryCodes
+
+  if (!isPublic) {
+    const { permission_groups: projectPermissionGroups } = await models.project.findOne({
+      code: project.code,
+    });
+
+    // Fetch all country codes we have any of the project permission groups access to
+    const projectAccessibleCountries: string[] = [];
+
+    for (const permission of projectPermissionGroups) {
+      projectAccessibleCountries.push(...(accessPolicy.getEntitiesAllowed(permission) || []));
+    }
+    allowedCountries = allowedCountries.filter(countryCode =>
+      projectAccessibleCountries.includes(countryCode),
+    );
+  }
+
+  return allowedCountries;
+};
+
+const getRecentEntities = async (
+  models: DatatrakWebModelRegistry,
+  user: CurrentUser,
+  countryCode: string,
+  type: string,
+  entities: EntityRecord[],
+) => {
+  // For public surveys
+  if (user.isLoggedIn) {
+    const recentEntities = await models.user.getRecentEntities(
+      user.id,
+      countryCode as string,
+      type as string,
+    );
+    return recentEntities
+      .map((id: string) => {
+        const entity = entities.find((e: EntityRecord) => e.id === id);
+        if (!entity) return null;
+        return { ...entity, isRecent: true };
+      })
+      .filter(Boolean);
```

Comment:
Unfortunately `Boolean.constructor` doesn’t act as a type guard, so TS will think that the return type is `(Entity | null)[]`

Using `isNotNullish` here will tighten that up

```suggestion
      .filter(isNotNullish);
```